File: apps/desktop/src/features/character/characterPresentation.ts
import type { CharacterPresentationSettingsDto } from '@parallel-world/contracts';
import type { CharacterController, CharacterModelSource } from '@parallel-world/live2d-runtime';

export const CHARACTER_PRESENTATION_SCHEMA_VERSION = 1;
export const CHARACTER_PRESENTATION_CHANGED_EVENT = 'character-presentation://changed';

export interface CharacterPresentationTransport {
  get(): Promise<unknown>;
  listen(listener: (value: unknown) => void): Promise<() => void>;
}

export function isCurrentCharacterPresentation(value: unknown): value is CharacterPresentationSettingsDto {
  if (!value || typeof value !== 'object') return false;
  const candidate = value as Record<string, unknown>;
  const shapeValid = candidate.schema_version === CHARACTER_PRESENTATION_SCHEMA_VERSION
    && Number.isSafeInteger(candidate.revision) && (candidate.revision as number) >= 0
    && typeof candidate.model_id === 'string' && typeof candidate.expression_id === 'string'
    && typeof candidate.motion_group === 'string' && Number.isInteger(candidate.motion_index)
    && typeof candidate.click_through === 'boolean';
  if (!shapeValid) return false;
  const model = candidate.model_id as string;
  const expression = candidate.expression_id as string;
  const group = candidate.motion_group as string;
  const index = candidate.motion_index as number;
  const expressionValid = model === 'mark' ? expression === '' : model === 'epsilon-free' && ['Angry', 'Blushing', 'f01', 'f02', 'Normal', 'Sad', 'Smile', 'Surprised'].includes(expression);
  const limit = model === 'mark' && group === 'Idle' ? 6 : model === 'epsilon-free' ? ({ Idle: 1, FlickUp: 2, Flick: 2, Tap: 4, Flick3: 2, FlickDown: 2, Shake: 2 } as Record<string, number>)[group] : undefined;
  return expressionValid && limit !== undefined && index >= 0 && index < limit;
}

export type CharacterModelSourceResolver = (modelId: string) => CharacterModelSource;
export const characterModelSource: CharacterModelSourceResolver = modelId => ({ modelId, manifestUrl: `characters/${modelId}/character.json` });

export function bindCharacterPresentation(controller: CharacterController, transport: CharacterPresentationTransport, resolveModelSource: CharacterModelSourceResolver = characterModelSource) {
  let active = true;
  let currentModelId: string | undefined;
  let scheduledRevision = -1;
  let unlisten: (() => void) | undefined;
  let pending = Promise.resolve();
  let failure: unknown;
  const removeListener = () => { const remove = unlisten; unlisten = undefined; remove?.(); };
  const enqueue = (unknownValue: unknown) => {
    if (!isCurrentCharacterPresentation(unknownValue)) return;
    if (unknownValue.revision <= scheduledRevision) return;
    scheduledRevision = unknownValue.revision;
    pending = pending.then(async () => {
      if (!active) return;
      if (unknownValue.model_id !== currentModelId) {
        await controller.loadModel(resolveModelSource(unknownValue.model_id));
        currentModelId = unknownValue.model_id;
      }
      if (unknownValue.expression_id) await controller.setExpression(unknownValue.expression_id);
      await controller.playMotion(unknownValue.motion_group, unknownValue.motion_index);
    }).catch(error => { failure = error; removeListener(); });
  };
  const ready = (async () => {
    try {
      const registeredUnlisten = await transport.listen(enqueue);
      if (!active) { registeredUnlisten(); return; }
      unlisten = registeredUnlisten;
      enqueue(await transport.get());
      await pending;
      if (failure) throw failure;
    } catch (error) {
      removeListener();
      throw error;
    }
  })();
  return {
    ready,
    flush: async () => { await pending; if (failure) throw failure; },
    dispose: () => { active = false; removeListener(); },
  };
}

File: crates/pw-domain/src/character_presentation.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CharacterPresentationSettings {
    model_id: String,
    expression_id: String,
    motion_group: String,
    motion_index: u32,
    click_through: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CharacterPresentationValidationError {
    UnknownModel,
    UnknownExpression,
    UnknownMotion,
}

impl CharacterPresentationSettings {
    pub fn try_new(
        model_id: &str,
        expression_id: &str,
        motion_group: &str,
        motion_index: u32,
        click_through: bool,
    ) -> Result<Self, CharacterPresentationValidationError> {
        let valid_expression = match model_id {
            "mark" => expression_id.is_empty(),
            "epsilon-free" => [
                "Angry",
                "Blushing",
                "f01",
                "f02",
                "Normal",
                "Sad",
                "Smile",
                "Surprised",
            ]
            .contains(&expression_id),
            _ => return Err(CharacterPresentationValidationError::UnknownModel),
        };
        if !valid_expression {
            return Err(CharacterPresentationValidationError::UnknownExpression);
        }
        let limit = match (model_id, motion_group) {
            ("mark", "Idle") => 6,
            ("epsilon-free", "Idle") => 1,
            ("epsilon-free", "FlickUp" | "Flick" | "Flick3" | "FlickDown" | "Shake") => 2,
            ("epsilon-free", "Tap") => 4,
            _ => return Err(CharacterPresentationValidationError::UnknownMotion),
        };
        if motion_index >= limit {
            return Err(CharacterPresentationValidationError::UnknownMotion);
        }
        Ok(Self {
            model_id: model_id.into(),
            expression_id: expression_id.into(),
            motion_group: motion_group.into(),
            motion_index,
            click_through,
        })
    }
    pub fn model_id(&self) -> &str {
        &self.model_id
    }
    pub fn expression_id(&self) -> &str {
        &self.expression_id
    }
    pub fn motion_group(&self) -> &str {
        &self.motion_group
    }
    pub fn motion_index(&self) -> u32 {
        self.motion_index
    }
    pub fn click_through(&self) -> bool {
        self.click_through
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accepts_phase_one_catalog_entries() {
        assert!(
            CharacterPresentationSettings::try_new("epsilon-free", "Smile", "Tap", 3, false)
                .is_ok()
        );
        assert!(CharacterPresentationSettings::try_new("mark", "", "Idle", 5, true).is_ok());
    }

    #[test]
    fn rejects_unknown_model_expression_and_motion() {
        assert_eq!(
            CharacterPresentationSettings::try_new("other", "", "Idle", 0, false),
            Err(CharacterPresentationValidationError::UnknownModel)
        );
        assert_eq!(
            CharacterPresentationSettings::try_new("epsilon-free", "Nope", "Idle", 0, false),
            Err(CharacterPresentationValidationError::UnknownExpression)
        );
        assert_eq!(
            CharacterPresentationSettings::try_new("epsilon-free", "Smile", "Nope", 0, false),
            Err(CharacterPresentationValidationError::UnknownMotion)
        );
        assert_eq!(
            CharacterPresentationSettings::try_new("epsilon-free", "Smile", "Tap", 4, false),
            Err(CharacterPresentationValidationError::UnknownMotion)
        );
    }
}
